src/import-data/ImportData.test.ts

```diff
@@ -0,0 +1,448 @@
+    useWorkspaces: jest.fn(),
+  };
+});
+
+type AjaxExports = typeof import('src/libs/ajax');
+type AjaxContract = ReturnType<AjaxExports['Ajax']>;
+jest.mock('src/libs/ajax');
+
+type NavExports = typeof import('src/libs/nav');
+jest.mock('src/libs/nav', (): NavExports => {
+  return {
+    ...jest.requireActual<NavExports>('src/libs/nav'),
+    goToPath: jest.fn(),
+    useRoute: jest.fn(),
+  };
+});
+
+type NotificationsExports = typeof import('src/libs/notifications');
+jest.mock('src/libs/notifications', (): NotificationsExports => {
+  return {
+    ...jest.requireActual<NotificationsExports>('src/libs/notifications'),
+    notify: jest.fn(),
+  };
+});
+
+type DataBrowserUtilsExports = typeof import('src/pages/library/dataBrowser-utils');
+jest.mock('src/pages/library/dataBrowser-utils', (): DataBrowserUtilsExports => {
+  return {
+    ...jest.requireActual<DataBrowserUtilsExports>('src/pages/library/dataBrowser-utils'),
+    useDataCatalog: jest.fn(),
+  };
+});
+
+interface SetupOptions {
+  mockAjax: DeepPartial<AjaxContract>;
+  queryParams: { [key: string]: unknown };
+}
+
+const setup = (opts: SetupOptions) => {
+  const { mockAjax, queryParams } = opts;
+
+  const mockAjaxWithCommonFunctions: DeepPartial<AjaxContract> = {
+    ...mockAjax,
+    Billing: {
+      listProjects: jest.fn().mockResolvedValue([{}]),
+      ...mockAjax.Billing,
+    },
+    FirecloudBucket: {
+      getTemplateWorkspaces: jest.fn().mockResolvedValue([]),
+      ...mockAjax.FirecloudBucket,
+    },
+    Metrics: {
+      captureEvent: jest.fn(),
+      ...mockAjax.Metrics,
+    },
+  };
+  asMockedFn(Ajax).mockImplementation(() => mockAjaxWithCommonFunctions as AjaxContract);
+
+  asMockedFn(useRoute).mockReturnValue({
+    query: queryParams,
+  });
+
+  render(h(ImportData));
+};
+
+const importIntoExistingWorkspace = async (user: UserEvent, workspaceName: string): Promise<void> => {
+  const existingWorkspace = screen.getByText('Start with an existing workspace', { exact: false });
+  await user.click(existingWorkspace);
+
+  const workspaceSelect = new SelectHelper(screen.getByLabelText('Select a workspace'), user);
+  await workspaceSelect.selectOption(workspaceName);
+
+  await user.click(screen.getByRole('button', { name: 'Import' }));
+};
+
+describe('ImportData', () => {
+  beforeEach(() => {
+    // Arrange
+    asMockedFn(useWorkspaces).mockReturnValue({
+      workspaces: [defaultAzureWorkspace, defaultGoogleWorkspace],
+      loading: false,
+      refresh: () => Promise.resolve(),
+    });
+
+    asMockedFn(useDataCatalog).mockReturnValue({
+      dataCatalog: [],
+      loading: false,
+      refresh: () => Promise.resolve(),
+    });
+  });
+
+  describe('PFB imports', () => {
+    it('imports PFB files', async () => {
+      // Arrange
+      const user = userEvent.setup();
+
+      const importJob = jest.fn().mockResolvedValue({ jobId: 'new-job' });
+      const getWorkspaceApi = jest.fn().mockReturnValue({
```

Comment:
Instead of declaring `getWorkspaceApi` here, can you declare inside the `setup` and return it as a destructured param from the `setup` method?  Then, tests that _need_ it can choose to read it and add additional mocks on it, but tests that don't won't need to redeclare it.

I think this would also let you drop the extra mockAjax config throughout:

```
          Workspaces: {
            workspace: getWorkspaceApi,
          },
```

Reply:
My thinking in the way I organized it was that each test mocks the Ajax methods that it needs instead of having `setup` mock everything that's needed for every test.

Reply:
Similarly, perhaps with `importJob` and `importBagit`, can these all be declared inside `setup`, and then you can access them out here only when needed, either as top level destructured returned objects, or maybe via `getWorkspaceApi`, eg:


```
// arrange
const {getWorkspaceApi} = setup( ... )
// fiddle with getWorkspaceApi before acting
when(getWorkspaceApi)...then

// act
...

// assert
expect(getWorkspaceApi.importJob).toHaveBeenCalledWith(...)

```

Reply:
oops I piled on another comment before seeing your reply:

> My thinking in the way I organized it was that each test mocks the Ajax methods that it needs instead of having `setup` mock everything that's needed for every test.

I completely agree that each test should have explicit mocking for each thing that matters only to that test (and ideally, nothing else), but I think it's not entirely unreasonable for the `setup` to provide default stubbed behavior for dependencies that essentially should work in all cases, and saving on some of the boilerplate might call attention to the parts that matter most for the unit test.  Totally your call though!

Reply:
> saving on some of the boilerplate might call attention to the parts that matter most for the unit test

This was compelling enough to try it in https://github.com/DataBiosphere/terra-ui/pull/4293/commits/c25220fe646391b3e8bc0ca54b898d43aeadb60f. It did cut ~50 lines of code.

Reply:
Nice!!